### PR TITLE
Fix Optimistic Effects Handling

### PR DESCRIPTION
**Pull request by djhi**
- [x] Fixes optimistic effects handling
Before this fix, if you had a `<List>` with a `<DeleteButton undoable>` on every line, deleting multiples records one after the other would end up with an invalid state. Removed items would still appear in the list even if the server call was made.

- [x] Fixes `beforeunload` event handler removal

### CODE REVIEW

File: packages/ra-core/src/dataProvider/useDataProvider.ts [hunk 1]
```diff
@@ -386,13 +400,27 @@ const warnBeforeClosingWindow = event => {
...
-        )
-    );
-    optimisticCalls.splice(0, optimisticCalls.length);
+const replayOptimisticCalls = async () => {
+    let clone;
+
+    // We must perform any undoable queries first so that the effects of previous undoable
+    // queries do not conflict with this one.
+
+    // We only handle all sideeffects queries if there are no more undoable queries
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/useDataProvider.ts:409]:
```suggestion
    // We only handle all side effect queries if there are no more undoable queries
```

File: packages/ra-core/src/dataProvider/useDataProvider.ts [hunk 2]
```diff
@@ -386,13 +400,27 @@ const warnBeforeClosingWindow = event => {
...
+    if (undoableOptimisticCalls.length > 0) {
+        clone = [...undoableOptimisticCalls];
+        await Promise.all(
+            clone.map(params => Promise.resolve(doQuery.call(null, params)))
+        );
+        undoableOptimisticCalls.splice(0, undoableOptimisticCalls.length);
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/useDataProvider.ts:415]: This one should move just after the clone

File: packages/ra-core/src/dataProvider/useDataProvider.ts [hunk 3]
```diff
@@ -300,7 +305,10 @@ const performUndoableQuery = ({
             if (window) {
                 window.removeEventListener(
                     'beforeunload',
-                    warnBeforeClosingWindow
+                    warnBeforeClosingWindow,
+                    {
+                        capture: true,
+                    }
```
Comment (djhi) [inline packages/ra-core/src/dataProvider/useDataProvider.ts:311]: https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener#Matching_event_listeners_for_removal

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Please describe the problem this PR solves

**Review by fzaninotto — APPROVED** [review]

**Pull request merged**